posts/2022-01-03-kryo-pitfalls.adoc

```diff
@@ -0,0 +1,289 @@
+handled carefully.
+
+=== Registration
+
+While working with Kryo, we can _optionally_ register the classes that we want to serialize:
+
+[source,scala]
+----
+case class Person(id: Long, name: String, age: Int)
+val alice = Person(1L, "Alice", 42)
+
+val kryo = new Kryo()
+kryo.register(classOf[Person])
+
+// Write
+val output = new Output(new FileOutputStream("file.bin"))
+kryo.writeObject(output, alice)
+output.close()
+
+// Read
+val input = new Input(new FileInputStream("file.bin"))
+val aliceFromFile = kryo.readObject(input, classOf[Person])
+input.close()
+----
+
+*Note*: to make the previous snippet work, you have to add a _no-args_ constructor to `Person`. I decided
+to omit it for brevity, but it's required for every class that you serialize with Kryo.
+
+As mentioned, class registration is optional, but required by default, which can be disabled with
+`kryo.setRegistrationRequired(false)`.
+
+If we analyze the output from both procedures we find them identical:
+
+[source,shell]
+----
+❯ xxd object-with-registration.bin
+00000000: 0154 0201 416c 6963 e5                   .T..Alic.
+
+❯ xxd object-without-registration.bin
+00000000: 0154 0201 416c 6963 e5                   .T..Alic.
+----
+
+== Pitfall #1: Compatibility
+
+Initially one could be led to believe that there is no difference between registering and not registering
+a class, but let's take a look at the next snippet:
+
+[source,scala]
+----
+case class Message(wholePayload: String, size: Int, ts: Long)  // <- New Message class
+case class Person(id: Long, name: String, age: Int)
+val alice = Person(1L, "Alice", 42)
+
+val kryo = new Kryo()
+kryo.register(classOf[Person])
+kryo.register(classOf[Message])
+
+val output = new Output(new FileOutputStream("file.bin"))
+kryo.writeObject(output, alice)                                // <--- Write a Person
+output.close()
+
+val input = new Input(new FileInputStream("file.bin"))
+val messageFromFile = kryo.readObject(input, classOf[Message]) // <--- Read a Message
+input.close()
+----
+
+This is perfectly valid, even if `Message` and `Person` are completely different, they have different field
+names. The only similarity is that both classes have three fields with the same types, just with a different
+order.
+The reason why this works has to do with how `FieldSerializer` (the default serializer for objects) is
+implemented:
+
+____
+FieldSerializer is efficient by writing only the field data, without any schema information, using the
+Java class files as the schema. It does not support adding, removing, or changing the type of fields
+without invalidating previously serialized bytes. Renaming fields is allowed only if it doesn't change
+the alphabetical order of the fields.
+____
+
+Since both classes' fields, once sorted by name, define the same order:
+`age, id, name` `==` `size, timestamp, wholePayload`,  Kryo deserializes a class' data into another class.
```

Comment:
I wonder if it wouldn't be better to add the types of the fields to avoid having the reader having to sort them in "their head". Like:

```suggestion
`age: Int, id: Long, name: String` `==` `size: Int, timestamp: Long, wholePayload: String`,  Kryo deserializes a class' data into another class.
```

Also note that the naming of the field `timestamp` is inconsistent with the name `ts` in the code sample.

Reply:
That makes sense, actually I'll tweak this sentence a bit to make the example clearer 👍